src/ball_sort/puzzle.cpp

```diff
@@ -0,0 +1,186 @@
+}
+
+auto Puzzle::make_tubes(const std::string &number_string) -> std::vector<Tube>
+{
+    std::vector<Tube> tubes{};
+
+    std::stringstream number_stream{number_string};
+    int number{};
+
+    std::string ball_string{};
+
+    while (number_stream >> number) {
+        ball_string.push_back(static_cast<char>('A' - 1 + number));
+        if (ball_string.size() == 4) {
+            tubes.emplace_back(ball_string);
+            ball_string.clear();
+        }
+    }
+
+    tubes.emplace_back("");
+    tubes.emplace_back("");
+
+    return tubes;
+}
+
+auto Puzzle::validate_puzzle() const -> void
+{
+    std::unordered_map<char, size_t> ball_tally{get_ball_tally()};
+
+    bool is_valid_puzzle{true};
+
+    for (const auto &tally : ball_tally) {
+        bool is_wrong_ball_quantity{tally.second != Tube::MAX_CAPACITY};
+        if (is_wrong_ball_quantity) is_valid_puzzle = false;
+    }
+
+    size_t minimum_number_of_tubes{ball_tally.size() + 2};
+    if (m_tubes.size() < minimum_number_of_tubes) is_valid_puzzle = false;
+
+    if (!is_valid_puzzle) {
+        std::cout << "Invalid puzzle" << '\n';
+        exit(1);
+    }
+}
+
+auto Puzzle::get_legal_moves() const -> std::vector<Move>
+{
+    std::vector<Move> legal_moves{};
+    const std::string &current_state = get_serialised_state();
+
+    // Iterates over every combination of origin and destination tube indices
+    for (size_t o{0}; o < m_tubes.size(); ++o) {
+        for (size_t d{0}; d < m_tubes.size(); ++d) {
+            if (is_legal_move(o, d))
+                legal_moves.emplace_back(o, d, current_state);
+        }
+    }
+
+    return legal_moves;
+}
+
+auto Puzzle::do_move(const size_t origin, const size_t destination) -> void
+{
+    const std::string &state_prior_to_move{get_serialised_state()};
+    m_move_history.emplace_back(origin, destination, state_prior_to_move);
+
+    char ball{m_tubes.at(origin).take_top_ball()};
+    m_tubes.at(destination).place_ball(ball);
+
+    const std::string &current_state{get_serialised_state()};
+
+    if (m_visited_puzzle_states.contains(current_state)) {
+        m_is_novel_puzzle_state = false;
+    } else {
+        m_visited_puzzle_states.insert(current_state);
+        m_is_novel_puzzle_state = true;
+    }
+}
+
+auto Puzzle::undo_move() -> void
+{
+    const Move &last_move{m_move_history.back()};
+
+    char ball{m_tubes.at(last_move.m_destination).take_top_ball()};
+    m_tubes.at(last_move.m_origin).place_ball(ball);
+
+    m_is_novel_puzzle_state = false;
+    m_move_history.pop_back();
+}
+
+auto Puzzle::is_novel_puzzle_state() const -> bool
+{
+    return m_is_novel_puzzle_state;
+}
+
+auto Puzzle::is_solved() const -> bool
+{
+    for (const Tube &tube : m_tubes) {
+        if (tube.is_empty()) continue;
+        if (!tube.is_solved()) return false;
+    }
+
+    return true;
+}
+
+auto Puzzle::reset() -> void
+{
+    m_tubes = m_initial_state;
+    m_move_history.clear();
+    m_visited_puzzle_states.clear();
+}
```

Comment:
When calling solver::play_solution, the history (representing the solution - perhaps I should have called it that) is copied out and the puzzle is reset prior to playback, so as to not create a duplicate history (which would affect sequential playbacks). So the instance is reused.

It could probably do with a rethink, but presently I'm just trying to stay afloat in a sea of new things to learn whilst implementing as many of your suggestions as I can manage :]

I might rename that variable, though.